### ISSUE TITLE
Remove deprecated CommonJS Node.js API

### ISSUE THREAD

**Issue by ybiquitous** [stylelint/stylelint#7395]
When releasing 16.0.0, we announced we would remove the CommonJS Node.js API in the next major version. See the migration guide:

https://github.com/stylelint/stylelint/blob/da7ce21272d34928f40bbc68c1e01be1c076bb8a/docs/migration-guide/to-16.md?plain=1#L5-L7

Also, we must drop the support for CommonJS plugins because Stylelint plugins written in CommonJS cannot import pure ESM Stylelint modules (e.g., `require('stylelint')` will be errored).

https://github.com/stylelint/stylelint/blob/da7ce21272d34928f40bbc68c1e01be1c076bb8a/lib/augmentConfig.mjs#L323-L328

By doing this, we will be able to achieve:

- Remove `.cjs` files (the package size reduces)
- Remove the build workflow via Rollup
- ~~Make the codebase pure ESM (maybe with `.mjs` → `.js` change?)~~
- ~~Bump outdated pure ESM dependencies~~

**Comment by ybiquitous** [comment]
@jeddy3 If you have additional info or find some mistakes, please feel free to edit the issue description.

**Comment by jeddy3** [comment]
All looks good to me! I've created a new label for things that are agreed on but held for a time.

**Comment by ybiquitous** [comment]
Yeah, it will be a while (maybe 1 year later?)

**Comment by ybiquitous** [comment]
I've just opened an issue #7397, too.

**Comment by jeddy3** [comment]
> Yeah, it will be a while (maybe 1 year later?)

I agree that it may be a while. We can check every few months to see how far along the JavaScript and Stylelint communities are in migrating to ESM, rather than put a fixed time on it.

**Comment by hudochenkov** [comment]
[require(esm) Backported to Node.js 20](https://socket.dev/blog/require-esm-backported-to-node-js-20):

> This change means that when Node.js 18 reaches end-of-life (EOL) in April 2025, library maintainers can finally drop their CJS builds and ship **ESM-only** packages with confidence.

**Comment by jeddy3** [comment]
SGTM. We can shoot for a 17.0.0 release at the end of April.

**Comment by ybiquitous** [comment]
Closed via #8859
